### ISSUE TITLE
QR code Scanning doesn't work on some android phones

### ISSUE THREAD

**Issue by trampster** [Redth/ZXing.Net.Maui#200]
QR code scanning doesn't work on the following android phones:

- Motorola G84 5G
- Sonim XP9900
- Oppo A54

The camera view shows but it never detects the QR code.

The preloaded camera app on these phones can scan these code successfully. And the same QR code can be scanned by other phones models.

![image](https://github.com/user-attachments/assets/bf562738-5178-4990-8086-2dbd14827ced)

I have attached logs from trying to scan with the Oppo A54. There are logs from when the scan page is loaded but there is nothing apart from GC collection stats while its actually trying to scan.

[OPPO A54 QR code scan logs.txt](https://github.com/user-attachments/files/17053053/OPPO.A54.QR.code.scan.logs.txt)

We are using:
- ZXing.Net.Maui version 0.4.0
- ZXing.Net.Maui.Controls version 0.4.0
- net8.0
- Maui version 8.0.80

**Comment by trampster** [comment]
After doing some debugging in ZXing.Net.Maui we have discovered that on the phones that don't work a comera frame is being retrieved and sent to ZXing.Net which is returning null.

I suspect that there is something wrong with the luminance data.

I have managed to dump the raw data from PixelBufferHolder.Data from working and not working phones.

Interestingly the data on the phone that isn't working is significantly larger, 491 kB instead of 307 kB . Given that both should be 640x480 then they should be exactly the same size.

I've attached the raw YUV data from both the working and the not working phones.

[WorkingYUVData.zip](https://github.com/user-attachments/files/17067833/WorkingYUVData.zip)
[NotWorkingYUVData.zip](https://github.com/user-attachments/files/17067834/NotWorkingYUVData.zip)

**Comment by thdigalakis** [comment]
Hi @trampster! Do you find any workaround for this issue?

**Comment by trampster** [comment]
No, we have just finished switching to [BarcodeScanning.Native.Maui](https://github.com/afriscic/BarcodeScanning.Native.Maui). It uses native scanning on each platform, and it works on all the android and iPhones we have tested on so far, including the ones that fail with ZXing.Net.Maui.

**Comment by alex-kir** [comment]
Hi,
I created a pull_request to fix a similar issue
https://github.com/Redth/ZXing.Net.Maui/pull/202
Now it works like a charm for me.

Using platform decoders does not work for Amazon devices.

**Comment by trampster** [comment]
Have you tried  BarcodeScanning.Native.Maui on amazon devices, it doesn't require Google Play Services (I've tested with it disabled, as it bundles the code with your app). But I don't have access to an amazon device.

We have also found BarcodeScanning.Native.Maui to be a lot faster, if you look at the logs while ZXing.Net.Maui is running it causes a Garbage collection every few seconds because of the massive amounts of memory allocations it does for every single frame, I see that your merge request adds even more memory allocations.

**Comment by alex-kir** [comment]
I didn't try  BarcodeScanning.Native.Maui but for several years I use my own solution based on `Android.Gms.Vision.Barcodes.BarcodeDetector` and on amazon devices `detector.IsOperational` always returns `false`.
As I understand the `Google.MLKit.Vision.BarCode.BarcodeScanning` is just a new API for the same functionality.

Now I believe that MLKit will not work too on Amazon devices. Here is an indirect confirmation (maybe in the future I will try to get direct confirmation, but now I don't have enough time)
https://developers.google.com/ml-kit/mobile-vision-migration/android#gmv
GMV uses the isOperational() call in the API surface to indicate whether a module has been downloaded successfully and is ready to use. ML Kit has no such method. ML Kit throws an MlKitException.UNAVAILABLE exception if a module hasn't been downloaded. You can catch this exception and process the next frame or set a timeout and retry with the current frame.

In my merge request tempBuffer allocated only once (if need) and reused in the next frames without any GC triggering.
I understand that the better solution is propagate `stride` outside but in this case, changes are needed in many places of the project, and can also affect user code

**Comment by kofanov** [comment]
Same problem with QR code scanning on Honor x9a (Android 13).

**Comment by OlivierBouchoms** [comment]
> Hi, I created a pull_request to fix a similar issue [#202](https://github.com/Redth/ZXing.Net.Maui/pull/202) Now it works like a charm for me.
> 
> Using platform decoders does not work for Amazon devices.

Thanks, I was running into this problem using ZXing in native Android, when testing with a OnePlus Nord CE 3 Lite (CPH2465).

Implementing your solution worked for me! For future reference, here's my Kotlin implementation of a CameraX QR Code analyzer:


```kotlin
package com.example.analyzer

import android.graphics.ImageFormat
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.google.zxing.BinaryBitmap
import com.google.zxing.PlanarYUVLuminanceSource
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.multi.qrcode.QRCodeMultiReader
import java.nio.ByteBuffer

/**
 * Class for analyzing frames of camera to check for QR Code, and decode QR Code
 * Code is referenced from: https://learntodroid.com/how-to-create-a-qr-code-scanner-app-in-android/
 */
class QRCodeAnalyzer(private var isEnabled: Boolean, private val listener: QRCodeListener) :
    ImageAnalysis.Analyzer {
    private val supportedFormats =
        arrayOf(ImageFormat.YUV_420_888, ImageFormat.YUV_422_888, ImageFormat.YUV_444_888)

    private var tempBuffer: ByteBuffer = ByteBuffer.allocate(0)
    private var tempRow: ByteArray = ByteArray(0)


    override fun analyze(image: ImageProxy) {
        if (!isEnabled) {
            image.close()
            return
        }

        if (!supportedFormats.contains(image.format)) {
            image.close()
            return
        }

        val imageBuffer = getImageBuffer(image)
        val imageData = ByteArray(imageBuffer.capacity())
        imageBuffer.get(imageData)

        val imageWidth = image.width
        val imageHeight = image.height

        val source = PlanarYUVLuminanceSource(
            imageData, imageWidth, imageHeight,
            0, 0, imageWidth, imageHeight, false
        )

        val binaryBitmap = BinaryBitmap(HybridBinarizer(source))

        try {
            val result = QRCodeMultiReader().decode(binaryBitmap)
            listener.qrCodeFound(result.text)
        } catch (e: Exception) {
            // ignore, as it is generally expected that the QR can't be found
        } finally {
            image.close()
        }
    }

    fun getImageBuffer(image: ImageProxy): ByteBuffer {
        val plane = image.planes[0]
        val stride = plane.rowStride
        val width = image.width

        if (stride == width) {
            return plane.buffer
        }

        // workaround for some devices (e.g. OnePlus CPH2465) which require different processing

        if (tempRow.size < width) {
            tempRow = ByteArray(width)
        }

        val height = image.height
        val bufferSize: Int = width * height
        if (tempBuffer.capacity() < bufferSize) {
            tempBuffer = ByteBuffer.allocate(bufferSize)
        }

        val source = plane.buffer
        for (y in 0..<height) {
            source.position(y * stride)
            source.get(tempRow, 0, width)
            tempBuffer.position(y * width)
            tempBuffer.put(tempRow, 0, width)
        }

        tempBuffer.position(0)

        return tempBuffer
    }

    fun setEnabled(value: Boolean) {
        isEnabled = value
    }

    interface QRCodeListener {
        fun qrCodeFound(qr: String)
    }
}
```
